File: src/flujo/plano/trazador.py
"""Convierte una imagen (PNG/JPG) en el contorno vectorial de un simbolo.

Para que la jefa de eventos pueda subir un icono aunque no lo tenga en SVG
(pedido del usuario, 2026-07-26). El resultado se le MUESTRA antes de guardar:
un trazado automatico puede salir sucio y quien decide si sirve es ella, no el
programa.

Por que se traza y no se incrusta la imagen tal cual: el simbolo termina en un
A4 impreso que recibe el productor tecnico del recinto. Un PNG chico incrustado
se ve pixelado al imprimir; un contorno vectorial escala limpio a cualquier
tamano y ademas obedece el color que se le declare, igual que los iconos de
fabrica.

Alcance honesto: esto traza SILUETAS, que es lo que es un icono. No reproduce
degradados, ni fotos, ni medios tonos. Si le das una foto vas a obtener una
mancha, y por eso existe la vista previa.

Solo usa Pillow, que es la unica dependencia de imagen declarada en
pyproject.toml. numpy/opencv estarian mas a mano pero no viajan con el paquete.
"""
from __future__ import annotations

import io
from typing import Dict, List, Sequence, Tuple

Punto = Tuple[float, float]
Lazo = List[Punto]

# Lado maximo de la grilla que se analiza. Mas resolucion no mejora un icono y
# multiplica los puntos del contorno (y el peso del SVG que se imprime).
LADO_MAX = 256

# Tolerancia de simplificacion, en celdas de la grilla. Sin esto el contorno
# sale con un vertice por pixel: pesado y con el borde dentado.
TOLERANCIA = 0.75

# Superficie minima de un lazo para conservarlo, como fraccion del total. Filtra
# el ruido suelto (motas del JPG) sin comerse detalles reales del icono.
AREA_MINIMA = 0.0006


class TrazadoImposible(Exception):
    """La imagen no da un contorno utilizable; el motivo va en el mensaje."""


def _mascara(imagen, umbral: int | None) -> Tuple[List[List[bool]], int, int]:
    """Grilla booleana: True donde hay tinta.

    Si la imagen trae transparencia, la tinta es lo opaco -- es el caso normal
    de un icono exportado. Si no, es lo oscuro sobre lo claro.
    """
    from PIL import Image

    imagen = imagen.copy()
    ancho, alto = imagen.size
    if max(ancho, alto) > LADO_MAX:
        escala = LADO_MAX / max(ancho, alto)
        imagen = imagen.resize((max(1, int(ancho * escala)), max(1, int(alto * escala))),
                               Image.LANCZOS)
    ancho, alto = imagen.size

    if imagen.mode in ("RGBA", "LA") or "transparency" in imagen.info:
        alfa = imagen.convert("RGBA").getchannel("A")
        datos = list(alfa.getdata())
        corte = 128 if umbral is None else umbral
        plano = [v > corte for v in datos]
        if any(plano):
            return [plano[f * ancho:(f + 1) * ancho] for f in range(alto)], ancho, alto

    gris = imagen.convert("L")
    datos = list(gris.getdata())
    corte = _umbral_otsu(datos) if umbral is None else umbral
    plano = [v < corte for v in datos]
    # Si el icono es claro sobre fondo oscuro, lo anterior toma el fondo.
    if sum(plano) > len(plano) * 0.6:
        plano = [not v for v in plano]
    return [plano[f * ancho:(f + 1) * ancho] for f in range(alto)], ancho, alto


def _umbral_otsu(valores: Sequence[int]) -> int:
    """Corte automatico entre tinta y fondo (Otsu), sin numpy."""
    histo = [0] * 256
    for v in valores:
        histo[v] += 1
    total = len(valores)
    suma_total = sum(i * histo[i] for i in range(256))
    suma_fondo = 0.0
    peso_fondo = 0
    mejor_var = -1.0
    mejor_corte = 128
    for i in range(256):
        peso_fondo += histo[i]
        if peso_fondo == 0:
            continue
        peso_frente = total - peso_fondo
        if peso_frente == 0:
            break
        suma_fondo += i * histo[i]
        media_fondo = suma_fondo / peso_fondo
        media_frente = (suma_total - suma_fondo) / peso_frente
        var = peso_fondo * peso_frente * (media_fondo - media_frente) ** 2
        if var > mejor_var:
            mejor_var, mejor_corte = var, i
    return mejor_corte


# Marching squares: por cada celda de 2x2, que segmentos de borde emite. Las
# claves son el caso (esquinas ocupadas como bits) y los valores, pares de
# puntos medios de los lados: 0=arriba 1=derecha 2=abajo 3=izquierda.
_LADOS = {0: (0.5, 0.0), 1: (1.0, 0.5), 2: (0.5, 1.0), 3: (0.0, 0.5)}
_CASOS: Dict[int, List[Tuple[int, int]]] = {
    1: [(3, 0)], 2: [(0, 1)], 3: [(3, 1)], 4: [(1, 2)], 5: [(3, 2), (0, 1)],
    6: [(0, 2)], 7: [(3, 2)], 8: [(2, 3)], 9: [(2, 0)], 10: [(0, 3), (2, 1)],
    11: [(2, 1)], 12: [(1, 3)], 13: [(1, 0)], 14: [(0, 3)],
}


def _contornos(mascara: List[List[bool]], ancho: int, alto: int) -> List[Lazo]:
    """Lazos cerrados que bordean la tinta, por marching squares."""
    def ocupado(f: int, c: int) -> bool:
        return 0 <= f < alto and 0 <= c < ancho and mascara[f][c]

    segmentos: Dict[Punto, List[Punto]] = {}
    for f in range(-1, alto):
        for c in range(-1, ancho):
            caso = (
                (1 if ocupado(f, c) else 0)
                | (2 if ocupado(f, c + 1) else 0)
                | (4 if ocupado(f + 1, c + 1) else 0)
                | (8 if ocupado(f + 1, c) else 0)
            )
            for desde, hasta in _CASOS.get(caso, []):
                a = (c + _LADOS[desde][0], f + _LADOS[desde][1])
                b = (c + _LADOS[hasta][0], f + _LADOS[hasta][1])
                segmentos.setdefault(a, []).append(b)

    lazos: List[Lazo] = []
    while segmentos:
        inicio = next(iter(segmentos))
        lazo: Lazo = [inicio]
        actual = inicio
        while True:
            salidas = segmentos.get(actual)
            if not salidas:
                break
            siguiente = salidas.pop()
            if not salidas:
                segmentos.pop(actual, None)
            if siguiente == inicio:
                break
            lazo.append(siguiente)
            actual = siguiente
        if len(lazo) > 3:
            lazos.append(lazo)
    return lazos


def _area(lazo: Lazo) -> float:
    s = 0.0
    for i in range(len(lazo)):
        x1, y1 = lazo[i]
        x2, y2 = lazo[(i + 1) % len(lazo)]
        s += x1 * y2 - x2 * y1
    return abs(s) / 2.0


def _simplificar(puntos: Lazo, tol: float) -> Lazo:
    """Douglas-Peucker iterativo (recursivo desborda con contornos largos)."""
    if len(puntos) < 3:
        return puntos
    conservar = [False] * len(puntos)
    conservar[0] = conservar[-1] = True
    pila = [(0, len(puntos) - 1)]
    while pila:
        ini, fin = pila.pop()
        if fin <= ini + 1:
            continue
        x1, y1 = puntos[ini]
        x2, y2 = puntos[fin]
        dx, dy = x2 - x1, y2 - y1
        norma = (dx * dx + dy * dy) ** 0.5
        peor, peor_i = -1.0, ini
        for i in range(ini + 1, fin):
            px, py = puntos[i]
            if norma == 0:
                d = ((px - x1) ** 2 + (py - y1) ** 2) ** 0.5
            else:
                d = abs(dy * px - dx * py + x2 * y1 - y2 * x1) / norma
            if d > peor:
                peor, peor_i = d, i
        if peor > tol:
            conservar[peor_i] = True
            pila.append((ini, peor_i))
            pila.append((peor_i, fin))
    return [p for p, keep in zip(puntos, conservar) if keep]


def trazar(datos: bytes, umbral: int | None = None, lado: int = 160) -> str:
    """Imagen -> SVG del contorno, en un lienzo cuadrado de `lado`.

    El trazo usa `currentColor`, que es la convencion que ya entiende el resto
    del catalogo: asi el mismo archivo sirve para el plano oscuro y el blanco.
    """
    from PIL import Image, UnidentifiedImageError

    try:
        imagen = Image.open(io.BytesIO(datos))
        imagen.load()
    except UnidentifiedImageError:
        raise TrazadoImposible("Ese archivo no es una imagen que pueda leer.")
    except Exception as e:  # noqa: BLE001
        raise TrazadoImposible(f"No pude abrir la imagen ({e}).")

    mascara, ancho, alto = _mascara(imagen, umbral)
    tinta = sum(sum(fila) for fila in mascara)
    if tinta == 0:
        raise TrazadoImposible("La imagen salio vacia: probá con una de más contraste.")
    if tinta == ancho * alto:
        raise TrazadoImposible("La imagen salio toda llena: no tiene contraste.")

    lazos = _contornos(mascara, ancho, alto)
    if not lazos:
        raise TrazadoImposible("No encontré un contorno en esa imagen.")

    area_total = max(_area(l) for l in lazos)
    utiles = [l for l in lazos if _area(l) >= area_total * AREA_MINIMA]

    escala = lado / max(ancho, alto)
    dx = (lado - ancho * escala) / 2.0
    dy = (lado - alto * escala) / 2.0

    partes: List[str] = []
    for lazo in utiles:
        simple = _simplificar(lazo, TOLERANCIA)
        if len(simple) < 3:
            continue
        pts = [f"{x * escala + dx:.2f} {y * escala + dy:.2f}" for x, y in simple]
        partes.append("M " + " L ".join(pts) + " Z")
    if not partes:
        raise TrazadoImposible("El contorno quedó demasiado chico para usarlo.")

    # evenodd para que los huecos del icono queden calados y no rellenos.
    return (
        f'<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 {lado} {lado}">'
        f'<path d="{" ".join(partes)}" fill="currentColor" fill-rule="evenodd"/>'
        f'</svg>'
    )

File: web/src/data/trazador.ts
// trazador.ts -- imagen -> contorno SVG, DENTRO del navegador.
//
// Por que existe: el trazador bueno es Python (`src/flujo/plano/trazador.py`) y
// vive en el servidor. En el HTML suelto que se le manda a la encargada de
// eventos no hay servidor, asi que subir un PNG le devolvia un error. Y decirle
// "abri la app con py -m flujo app" es mandarla a una consola que no tiene:
// justamente por eso se le pasa un archivo y no un repositorio.
//
// Es el MISMO algoritmo, con las mismas constantes, para que un icono trazado
// aca y uno trazado alla no salgan distintos: umbral de Otsu, marching squares,
// y Douglas-Peucker iterativo (recursivo desborda con contornos largos).
// Si alguna vez se cambia uno, hay que cambiar el otro.

export class TrazadoImposible extends Error {}

const LADO_MAX = 256;
const TOLERANCIA = 0.75;
const AREA_MINIMA = 0.0006;

type Punto = [number, number];

/** Corte que mejor separa dos poblaciones de gris. Igual que `_umbral_otsu`. */
function umbralOtsu(hist: number[], total: number): number {
  let suma = 0;
  for (let i = 0; i < 256; i++) suma += i * hist[i];
  let sumaB = 0, pesoB = 0, mejorVar = -1, corte = 128;
  for (let t = 0; t < 256; t++) {
    pesoB += hist[t];
    if (pesoB === 0) continue;
    const pesoF = total - pesoB;
    if (pesoF === 0) break;
    sumaB += t * hist[t];
    const mediaB = sumaB / pesoB;
    const mediaF = (suma - sumaB) / pesoF;
    const entre = pesoB * pesoF * (mediaB - mediaF) * (mediaB - mediaF);
    if (entre > mejorVar) { mejorVar = entre; corte = t; }
  }
  return corte;
}

/** Grilla booleana: true donde hay tinta. */
function mascara(img: ImageData): { m: boolean[][]; ancho: number; alto: number } {
  const { width: ancho, height: alto, data } = img;
  const n = ancho * alto;

  // Con transparencia, la tinta es lo opaco: es el caso normal de un icono
  // exportado. Se prueba primero porque da un recorte mucho mas limpio.
  let plano: boolean[] = new Array(n);
  let conTinta = 0;
  for (let i = 0; i < n; i++) {
    const opaco = data[i * 4 + 3] > 128;
    plano[i] = opaco;
    if (opaco) conTinta++;
  }
  if (conTinta > 0 && conTinta < n) {
    return { m: aFilas(plano, ancho, alto), ancho, alto };
  }

  // Sin transparencia: lo oscuro sobre lo claro, con el corte de Otsu.
  const hist = new Array(256).fill(0);
  const gris = new Array(n);
  for (let i = 0; i < n; i++) {
    const g = Math.round(0.299 * data[i * 4] + 0.587 * data[i * 4 + 1]
                         + 0.114 * data[i * 4 + 2]);
    gris[i] = g;
    hist[g]++;
  }
  const corte = umbralOtsu(hist, n);
  plano = gris.map(g => g <= corte);
  // Si el icono es claro sobre fondo oscuro, lo anterior tomo el fondo.
  const tinta = plano.reduce((a, b) => a + (b ? 1 : 0), 0);
  if (tinta > n * 0.6) plano = plano.map(v => !v);
  return { m: aFilas(plano, ancho, alto), ancho, alto };
}

function aFilas(plano: boolean[], ancho: number, alto: number): boolean[][] {
  const filas: boolean[][] = [];
  for (let f = 0; f < alto; f++) filas.push(plano.slice(f * ancho, (f + 1) * ancho));
  return filas;
}

const LADOS: Record<number, [number, number]> = {
  0: [0.5, 0.0], 1: [1.0, 0.5], 2: [0.5, 1.0], 3: [0.0, 0.5],
};
const CASOS: Record<number, Array<[number, number]>> = {
  1: [[3, 0]], 2: [[0, 1]], 3: [[3, 1]], 4: [[1, 2]], 5: [[3, 2], [0, 1]],
  6: [[0, 2]], 7: [[3, 2]], 8: [[2, 3]], 9: [[2, 0]], 10: [[0, 3], [2, 1]],
  11: [[2, 1]], 12: [[1, 3]], 13: [[1, 0]], 14: [[0, 3]],
};

/** Lazos cerrados que bordean la tinta, por marching squares. */
function contornos(m: boolean[][], ancho: number, alto: number): Punto[][] {
  const ocupado = (f: number, c: number) =>
    f >= 0 && f < alto && c >= 0 && c < ancho && m[f][c];

  const segmentos = new Map<string, Punto[]>();
  const clave = (p: Punto) => `${p[0]},${p[1]}`;

  for (let f = -1; f < alto; f++) {
    for (let c = -1; c < ancho; c++) {
      const caso = (ocupado(f, c) ? 1 : 0) | (ocupado(f, c + 1) ? 2 : 0)
        | (ocupado(f + 1, c + 1) ? 4 : 0) | (ocupado(f + 1, c) ? 8 : 0);
      for (const [desde, hasta] of CASOS[caso] || []) {
        const a: Punto = [c + LADOS[desde][0], f + LADOS[desde][1]];
        const b: Punto = [c + LADOS[hasta][0], f + LADOS[hasta][1]];
        const k = clave(a);
        const lista = segmentos.get(k);
        if (lista) lista.push(b); else segmentos.set(k, [b]);
      }
    }
  }

  const puntoDe = (k: string): Punto => {
    const [x, y] = k.split(',');
    return [Number(x), Number(y)];
  };

  const lazos: Punto[][] = [];
  while (segmentos.size) {
    const kInicio = segmentos.keys().next().value as string;
    const inicio = puntoDe(kInicio);
    const lazo: Punto[] = [inicio];
    let actual = inicio;
    for (;;) {
      const salidas = segmentos.get(clave(actual));
      if (!salidas || !salidas.length) break;
      const siguiente = salidas.pop() as Punto;
      if (!salidas.length) segmentos.delete(clave(actual));
      if (clave(siguiente) === kInicio) break;
      lazo.push(siguiente);
      actual = siguiente;
    }
    if (lazo.length > 3) lazos.push(lazo);
  }
  return lazos;
}

function area(lazo: Punto[]): number {
  let s = 0;
  for (let i = 0; i < lazo.length; i++) {
    const [x1, y1] = lazo[i];
    const [x2, y2] = lazo[(i + 1) % lazo.length];
    s += x1 * y2 - x2 * y1;
  }
  return Math.abs(s) / 2;
}

/** Douglas-Peucker iterativo: el recursivo desborda con contornos largos. */
function simplificar(puntos: Punto[], tol: number): Punto[] {
  if (puntos.length < 3) return puntos;
  const conservar = new Array(puntos.length).fill(false);
  conservar[0] = conservar[puntos.length - 1] = true;
  const pila: Array<[number, number]> = [[0, puntos.length - 1]];
  while (pila.length) {
    const [ini, fin] = pila.pop() as [number, number];
    if (fin <= ini + 1) continue;
    const [x1, y1] = puntos[ini];
    const [x2, y2] = puntos[fin];
    const dx = x2 - x1, dy = y2 - y1;
    const norma = Math.hypot(dx, dy);
    let peor = -1, peorI = ini;
    for (let i = ini + 1; i < fin; i++) {
      const [px, py] = puntos[i];
      const d = norma === 0
        ? Math.hypot(px - x1, py - y1)
        : Math.abs(dy * px - dx * py + x2 * y1 - y2 * x1) / norma;
      if (d > peor) { peor = d; peorI = i; }
    }
    if (peor > tol) {
      conservar[peorI] = true;
      pila.push([ini, peorI], [peorI, fin]);
    }
  }
  return puntos.filter((_, i) => conservar[i]);
}

/** Lee la imagen a un lienzo, achicandola si hace falta. */
async function aImageData(archivo: File): Promise<ImageData> {
  const url = URL.createObjectURL(archivo);
  try {
    const img = await new Promise<HTMLImageElement>((ok, mal) => {
      const el = new Image();
      el.onload = () => ok(el);
      el.onerror = () => mal(new TrazadoImposible('Ese archivo no es una imagen que pueda leer.'));
      el.src = url;
    });
    let { naturalWidth: w, naturalHeight: h } = img;
    if (!w || !h) throw new TrazadoImposible('Esa imagen no tiene tamaño.');
    const escala = Math.min(1, LADO_MAX / Math.max(w, h));
    w = Math.max(1, Math.round(w * escala));
    h = Math.max(1, Math.round(h * escala));
    const lienzo = document.createElement('canvas');
    lienzo.width = w; lienzo.height = h;
    const ctx = lienzo.getContext('2d', { willReadFrequently: true });
    if (!ctx) throw new TrazadoImposible('El navegador no me dejó leer la imagen.');
    ctx.drawImage(img, 0, 0, w, h);
    return ctx.getImageData(0, 0, w, h);
  } finally {
    URL.revokeObjectURL(url);
  }
}

/** El nucleo, sobre pixeles crudos. Separado de la lectura del archivo para
 *  poder compararlo contra el trazador de Python con los MISMOS pixeles. */
export function trazarDesdePixeles(datos: ImageData, lado = 160): string {
  const { m, ancho, alto } = mascara(datos);

  let tinta = 0;
  for (const fila of m) for (const v of fila) if (v) tinta++;
  if (tinta === 0) {
    throw new TrazadoImposible('La imagen salió vacía: probá con una de más contraste.');
  }
  if (tinta === ancho * alto) {
    throw new TrazadoImposible('La imagen salió toda llena: no tiene contraste.');
  }

  const lazos = contornos(m, ancho, alto);
  if (!lazos.length) throw new TrazadoImposible('No encontré un contorno en esa imagen.');

  const mayor = Math.max(...lazos.map(area));
  const utiles = lazos.filter(l => area(l) >= mayor * AREA_MINIMA);

  const escala = lado / Math.max(ancho, alto);
  const dx = (lado - ancho * escala) / 2;
  const dy = (lado - alto * escala) / 2;

  const partes: string[] = [];
  for (const lazo of utiles) {
    const simple = simplificar(lazo, TOLERANCIA);
    if (simple.length < 3) continue;
    const pts = simple.map(([x, y]) =>
      `${(x * escala + dx).toFixed(2)} ${(y * escala + dy).toFixed(2)}`);
    partes.push(`M ${pts.join(' L ')} Z`);
  }
  if (!partes.length) {
    throw new TrazadoImposible('El contorno quedó demasiado chico para usarlo.');
  }

  // evenodd para que los huecos del icono queden calados y no rellenos.
  // currentColor es la convencion del resto del catalogo: el mismo archivo
  // sirve para el plano oscuro y para el blanco.
  return `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 ${lado} ${lado}">`
    + `<path d="${partes.join(' ')}" fill="currentColor" fill-rule="evenodd"/>`
    + `</svg>`;
}

/** Imagen -> SVG del contorno, en un lienzo cuadrado de `lado`. */
export async function trazarEnNavegador(archivo: File, lado = 160): Promise<string> {
  return trazarDesdePixeles(await aImageData(archivo), lado);
}
